File: Visualizer/src/main/java/com/github/ezauton/visualizer/processor/RobotStateDataProcessor.java
package com.github.ezauton.visualizer.processor;

import com.github.ezauton.recorder.base.RobotStateRecorder;
import com.github.ezauton.recorder.base.frame.RobotStateFrame;
import com.github.ezauton.visualizer.util.DataProcessor;
import com.github.ezauton.visualizer.util.Environment;
import javafx.animation.Interpolator;
import javafx.animation.KeyValue;
import javafx.scene.control.Label;
import javafx.scene.layout.GridPane;
import javafx.scene.paint.Paint;
import javafx.scene.shape.Circle;
import javafx.scene.shape.Rectangle;
import javafx.scene.transform.Rotate;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class RobotStateDataProcessor implements DataProcessor {
    private final RobotStateRecorder robotRec;
    private final List<RobotStateFrame> dataFrames;

    private Rectangle robot;
    private Circle posCircle;
    private Label headingLabel, posLabel, velocityLabel;
    private double spatialScaleFactorX;
    private double spatialScaleFactorY;
    private double originYPx;
    private double originXPx;
    private int robotLengthPx;
    private int robotWidthPx;

    public RobotStateDataProcessor(RobotStateRecorder robotRec) {
        this.robotRec = robotRec;
        dataFrames = robotRec.getDataFrames();
    }


    private double getX(double feetX) {
        return feetX * spatialScaleFactorX + originXPx;
    }

    private double getY(double feetY) {
        return -feetY * spatialScaleFactorY + originYPx;
    }

    @Override
    public void initEnvironment(Environment environment) {
        this.spatialScaleFactorX = environment.getScaleFactorX();
        this.spatialScaleFactorY = environment.getScaleFactorY();

        this.originXPx = environment.getOrigin().get(0);
        this.originYPx = environment.getOrigin().get(1);

        // box for robot
        robotWidthPx = (int) (dataFrames.get(0).getRobotWidth() * spatialScaleFactorX);
        robotLengthPx = (int) (dataFrames.get(0).getRobotLength() * spatialScaleFactorY);
        robot = new Rectangle(0, 0, robotWidthPx, robotLengthPx);
        posCircle = new Circle(3, Paint.valueOf("white"));
        posCircle.setStroke(Paint.valueOf("black"));


        // heading info
        headingLabel = new Label("0 radians");

        // x y loc
        posLabel = new Label("(0, 0)");

        velocityLabel = new Label("(0, 0)");

        robot.setFill(Paint.valueOf("cyan"));
        robot.setStroke(Paint.valueOf("black"));
        environment.getFieldAnchorPane().getChildren().add(0, robot);
        environment.getFieldAnchorPane().getChildren().add(posCircle);

        GridPane dataGridPane = environment.getDataGridPane(robotRec.getName());
        dataGridPane.addRow(0, new Label("Heading: "), headingLabel);
        dataGridPane.addRow(1, new Label("Position: "), posLabel);
        dataGridPane.addRow(2, new Label("Velocity: "), velocityLabel);
    }

    @Override
    public Map<Double, List<KeyValue>> generateKeyValues(Interpolator interpolator) {
        Map<Double, List<KeyValue>> ret = new HashMap<>();

        Rotate robotRotation = new Rotate();
        robot.getTransforms().add(robotRotation);

        for (RobotStateFrame frame : robotRec.getDataFrames()) {
            List<KeyValue> keyValues = new ArrayList<>();
            double x, y, heading;

            if (frame.getPos() != null) {
                x = getX(frame.getPos().get(0));
                y = getY(frame.getPos().get(1));

                heading = frame.getHeading();

                double robotX = x - robotWidthPx / 2D;
                double robotY = y - robotLengthPx;

                keyValues.add(new KeyValue(robot.xProperty(), robotX, interpolator));
                keyValues.add(new KeyValue(robot.yProperty(), robotY, interpolator));
//                keyValues.add(new KeyValue(robot.rotateProperty(), -Math.toDegrees(heading), interpolator));
                keyValues.add(new KeyValue(robotRotation.angleProperty(), -Math.toDegrees(heading), interpolator));
                keyValues.add(new KeyValue(robotRotation.pivotXProperty(), x, interpolator));
                keyValues.add(new KeyValue(robotRotation.pivotYProperty(), y, interpolator));
                keyValues.add(new KeyValue(robot.visibleProperty(), true, interpolator));
                keyValues.add(new KeyValue(posCircle.centerXProperty(), x, interpolator));
                keyValues.add(new KeyValue(posCircle.centerYProperty(), y, interpolator));
                keyValues.add(new KeyValue(headingLabel.textProperty(), String.format("%.02f radians", heading)));
                keyValues.add(new KeyValue(posLabel.textProperty(), String.format("(%.02f, %.02f)", frame.getPos().get(0), frame.getPos().get(1))));
                keyValues.add(new KeyValue(velocityLabel.textProperty(), String.format("(%.02f, %.02f)", frame.getRobotVelocity().get(0), frame.getRobotVelocity().get(1))));
            } else {
                keyValues.add(new KeyValue(robot.visibleProperty(), false, interpolator));
            }

            ret.put(frame.getTime(),
                    keyValues);
        }

        return ret;
    }
}

File: Visualizer/src/main/java/com/github/ezauton/visualizer/processor/RobotStateDataProcessor.kt
package com.github.ezauton.visualizer.processor

import com.github.ezauton.recorder.base.RobotStateRecording
import com.github.ezauton.recorder.base.frame.RobotStateFrame
import com.github.ezauton.visualizer.util.DataProcessor
import com.github.ezauton.visualizer.util.Environment
import javafx.animation.Interpolator
import javafx.animation.KeyValue
import javafx.scene.control.Label
import javafx.scene.layout.GridPane
import javafx.scene.paint.Paint
import javafx.scene.shape.Circle
import javafx.scene.shape.Rectangle
import javafx.scene.transform.Rotate

class RobotStateDataProcessor(private val robotRec: RobotStateRecording) : DataProcessor {
  private val dataFrames: List<RobotStateFrame> = robotRec.frames
  private lateinit var robot: Rectangle
  private lateinit var posCircle: Circle
  private lateinit var headingLabel: Label
  private lateinit var posLabel: Label
  private lateinit var velocityLabel: Label
  private var spatialScaleFactorX = 0.0
  private var spatialScaleFactorY = 0.0
  private var originYPx = 0.0
  private var originXPx = 0.0
  private var robotLengthPx = 0.0
  private var robotWidthPx = 0.0

  private fun getX(feetX: Double): Double {
    return feetX * spatialScaleFactorX + originXPx
  }

  private fun getY(feetY: Double): Double {
    return -feetY * spatialScaleFactorY + originYPx
  }

  override fun initEnvironment(environment: Environment) {
    spatialScaleFactorX = environment.scaleFactorX
    spatialScaleFactorY = environment.scaleFactorY
    originXPx = environment.origin[0]
    originYPx = environment.origin[1]

    // box for robot
    robotWidthPx = dataFrames[0].robotWidth * spatialScaleFactorX
    robotLengthPx = dataFrames[0].robotLength * spatialScaleFactorY
    robot = Rectangle(0.0, 0.0, robotWidthPx, robotLengthPx)
    posCircle = Circle(3.0, Paint.valueOf("white"))
    posCircle.stroke = Paint.valueOf("black")


    // heading info
    headingLabel = Label("0 radians")

    // x y loc
    posLabel = Label("(0, 0)")
    velocityLabel = Label("(0, 0)")
    robot.fill = Paint.valueOf("cyan")
    robot.stroke = Paint.valueOf("black")
    environment.fieldAnchorPane.children.add(0, robot)
    environment.fieldAnchorPane.children.add(posCircle)
    val dataGridPane: GridPane = environment.getDataGridPane(robotRec.name)
    dataGridPane.addRow(0, Label("Heading: "), headingLabel)
    dataGridPane.addRow(1, Label("Position: "), posLabel)
    dataGridPane.addRow(2, Label("Velocity: "), velocityLabel)
  }

  override fun generateKeyValues(interpolator: Interpolator): Map<Double, List<KeyValue>> {
    val ret: MutableMap<Double, MutableList<KeyValue>> = HashMap()
    val robotRotation = Rotate()
    robot.transforms.add(robotRotation)
    for (frame in robotRec.frames) {
      val keyValues: MutableList<KeyValue> = ArrayList()
      val x: Double = getX(frame.pos[0])
      val y: Double = getY(frame.pos[1])
      val heading: Double = frame.heading
      val robotX = x - robotWidthPx / 2.0
      val robotY = y - robotLengthPx
      keyValues.add(KeyValue(robot.xProperty(), robotX, interpolator))
      keyValues.add(KeyValue(robot.yProperty(), robotY, interpolator))
      //                keyValues.add(new KeyValue(robot.rotateProperty(), -Math.toDegrees(heading), interpolator));
      keyValues.add(KeyValue(robotRotation.angleProperty(), -Math.toDegrees(heading), interpolator))
      keyValues.add(KeyValue(robotRotation.pivotXProperty(), x, interpolator))
      keyValues.add(KeyValue(robotRotation.pivotYProperty(), y, interpolator))
      keyValues.add(KeyValue(robot.visibleProperty(), true, interpolator))
      keyValues.add(KeyValue(posCircle.centerXProperty(), x, interpolator))
      keyValues.add(KeyValue(posCircle.centerYProperty(), y, interpolator))
      keyValues.add(KeyValue(headingLabel.textProperty(), String.format("%.02f radians", heading)))
      keyValues.add(KeyValue(posLabel.textProperty(), String.format("(%.02f, %.02f)", frame.pos[0], frame.pos[1])))
      keyValues.add(KeyValue(velocityLabel.textProperty(), String.format("(%.02f, %.02f)", frame.robotVelocity[0], frame.robotVelocity[1])))
      ret[frame.time] = keyValues
    }
    return ret
  }

}
